sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala

```diff
@@ -0,0 +1,139 @@
+  /**
+   * Whether the statement originates from the SQL script or is created during the interpretation.
+   * Example: DropVariable statements are automatically created at the end of each compound.
+   */
+  val isInternal: Boolean = false
+
+  /**
+   * Reset execution of the current node.
+   */
+  def reset(): Unit
+}
+
+/**
+ * Leaf node in the execution tree.
+ */
+trait LeafStatementExec extends CompoundStatementExec
+
+/**
+ * Non-leaf node in the execution tree.
+ * It is an iterator over executable child nodes.
+ */
+trait NonLeafStatementExec extends CompoundStatementExec with Iterator[CompoundStatementExec]
+
+/**
+ * Executable node for SingleStatement.
+ * @param parsedPlan Logical plan of the parsed statement.
+ * @param origin Origin descriptor for the statement.
+ * @param isInternal Whether the statement originates from the SQL script or it is created during
+ *                   the interpretation. Example: DropVariable statements are automatically created
+ *                   at the end of each compound.
+ */
+class SingleStatementExec(
+    var parsedPlan: LogicalPlan,
+    override val origin: Origin,
+    override val isInternal : Boolean)
+  extends LeafStatementExec
+  with WithOrigin {
+
+  /**
+   * Whether this statement had to be executed during the interpretation phase.
+   * Example: Statements in conditions of If/Else, While, etc.
+   */
+  var consumed = false
+
+  /** @inheritdoc */
+  override def reset(): Unit = consumed = false
+
+  /** Get the SQL query text corresponding to this statement. */
+  def getText(sqlScriptText: String): String = {
+    if (origin.startIndex.isEmpty || origin.stopIndex.isEmpty) {
+      return null
+    }
+    sqlScriptText.substring(origin.startIndex.get, origin.stopIndex.get + 1)
+  }
+}
+
+/**
+ * Abstract class for all statements that contain nested statements.
+ * Implements recursive iterator logic over all child execution nodes.
+ * @param collection Collection of child execution nodes.
+ */
+abstract class CompoundNestedStatementIteratorExec(collection: Seq[CompoundStatementExec])
+  extends NonLeafStatementExec {
+
+  var localIterator = collection.iterator
+  var curr = if (localIterator.hasNext) Some(localIterator.next()) else None
+
+  override def hasNext: Boolean = {
+    val childHasNext = curr match {
+      case Some(body: NonLeafStatementExec) => body.hasNext
+      case Some(_: LeafStatementExec) => true
+      case None => false
+      case _ => throw SparkException.internalError(
+        "Unknown statement type encountered during SQL script interpretation.")
+    }
+    localIterator.hasNext || childHasNext
+  }
+
+  override def next(): CompoundStatementExec = {
+    curr match {
+      case None => throw SparkException.internalError(
+        "No more elements to iterate through in the current SQL compound statement.")
+      case Some(statement: LeafStatementExec) =>
+        if (localIterator.hasNext) curr = Some(localIterator.next())
```

Comment:
Could be rewritten in Scala fashion like below ( curr = if (localIterator...)

Reply:
literally have it few lines below the same way like you suggested... thanks!